Cabal/Distribution/Parsec/Parser.hs

```diff
@@ -0,0 +1,393 @@
+elementInNonLayoutContext name =
+      (do colon; fieldInlineOrBraces name)
+  <|> (do args <- many sectionArg
+          openBrace
+          elems <- elements zeroIndentLevel
+          optional tokIndent
+          closeBrace
+          return (Section name args elems))
+
+-- The body of a field, using either layout style or braces style.
+--
+-- fieldLayoutOrBraces   ::= '\n'? '{' content '}'
+--                         | line? ('\n' line)*
+fieldLayoutOrBraces :: IndentLevel -> Name Position -> Parser (Field Position)
+fieldLayoutOrBraces ilevel name =
+      (do openBrace
+          ls <- inLexerMode (LexerMode in_field_braces) (many fieldContent)
+          closeBrace
+          return (Field name ls))
+  <|> (inLexerMode (LexerMode in_field_layout)
+        (do l  <- option (FieldLine (Position 0 0) B8.empty) fieldContent
+                  --FIXME ^^ having to add an extra empty here is silly!
+            ls <- many (do _ <- indentOfAtLeast ilevel; fieldContent)
+            return (Field name (l:ls))))
+
+-- The body of a section, using either layout style or braces style.
+--
+-- sectionLayoutOrBraces ::= '\n'? '{' elements \n? '}'
+--                         | elements
+sectionLayoutOrBraces :: IndentLevel -> Parser [Field Position]
+sectionLayoutOrBraces ilevel =
+      (do openBrace
+          elems <- elements zeroIndentLevel
+          optional tokIndent
+          closeBrace
+          return elems)
+  <|> (elements ilevel)
+
+-- The body of a field, using either inline style or braces.
+--
+-- fieldInlineOrBraces   ::= '\n'? '{' content '}'
+--                         | content
+fieldInlineOrBraces :: Name Position -> Parser (Field Position)
+fieldInlineOrBraces name =
+      (do openBrace
+          ls <- inLexerMode (LexerMode in_field_braces) (many fieldContent)
+          closeBrace
+          return (Field name ls))
+  <|> (do ls <- inLexerMode (LexerMode in_field_braces) (option [] (fmap (\l -> [l]) fieldContent))
+          return (Field name ls))
+
+
+readFields :: B8.ByteString -> Either ParseError [Field Position]
+readFields s = fmap fst (readFields' s)
+
+readFields' :: B8.ByteString -> Either ParseError ([Field Position], [LexWarning])
+readFields' s = do
+    parse parser "the input" lexSt
+  where
+    parser = do
+        fields <- cabalStyleFile
+        ws     <- getLexerWarnings
+        pure (fields, maybeToList w ++ ws)
+
+    (w, s') = fmap B.pack . recodeStringUtf8 . B.unpack $ s
+    lexSt = mkLexState' (mkLexState s')
+
+-- TODO: For some reason alex parser cannot handle BOM, is it a bug?
+recodeStringUtf8 :: [Word8] -> (Maybe LexWarning, [Word8])
+recodeStringUtf8 (0xef : 0xbb : 0xbf : bytes) =
+    ( Just $ LexWarning LexWarningBOM (Position 1 1) "Byte-order mark found"
+    , encodeStringUtf8 (decodeStringUtf8 bytes)
+    )
+recodeStringUtf8 bytes =
+    (Nothing, encodeStringUtf8 (decodeStringUtf8 bytes))
+
+#ifdef CABAL_PARSEC_DEBUG
```

Comment:
IIRC it requires `dbgText` in lexer too. There was some complications. Have to recheck and comment about them,